File: capstone_launchpad/tests_backup/initializePool.ts
import * as anchor from "@coral-xyz/anchor";
import { Program } from "@coral-xyz/anchor";
import { CapstoneLaunchpad,  } from "../target/types/capstone_launchpad";
import { token } from "@coral-xyz/anchor/dist/cjs/utils";
import {
  ASSOCIATED_TOKEN_PROGRAM_ID,
  TOKEN_PROGRAM_ID,
  createMint,
  getAssociatedTokenAddressSync,
  getAccount,
  mintTo,
  createAssociatedTokenAccount,
  MINT_SIZE,
  createInitializeMintInstruction,
} from "@solana/spl-token";
import { PublicKey, SystemProgram, Transaction } from "@solana/web3.js";
import { AMMTestUtils, AMMTestSetup } from "../helper";

describe("capstone_launchpad", () => {
  // Configure the client to use the local cluster.
  const provider=anchor.AnchorProvider.env();
  const authority = provider.publicKey;

  anchor.setProvider(provider);
    const connection = provider.connection;
  const wallet = provider.wallet;

  const program = anchor.workspace.capstoneLaunchpad as Program<CapstoneLaunchpad>;
  const seed_string  = 5;

  it("Initializes a pool", async () => {
    const poolSeed = new anchor.BN(seed_string);
    const fee = 30; //0.3%
    const authority = provider.publicKey;

    const ammSeed = new anchor.BN(seed_string); 
    const [configAmmPda, configBump] = anchor.web3.PublicKey.findProgramAddressSync(
      [Buffer.from("config_amm"), ammSeed.toArrayLike(Buffer, "le", 8)],
      program.programId
    );

   const [configPoolPda, configPoolBump] = PublicKey.findProgramAddressSync(
     [Buffer.from("config_pool"), configAmmPda.toBuffer(), poolSeed.toArrayLike(Buffer, "le", 8)],
     program.programId
   );

    const [lpMintPda, lpBump] = anchor.web3.PublicKey.findProgramAddressSync(
      [Buffer.from("lp"), configPoolPda.toBuffer()],
      program.programId
    );

    console.log("Expected config_pool PDA:", configPoolPda.toString());
    console.log("Expected LP mint PDA:", lpMintPda.toString());

    // const mintX = await createMint(connection, wallet.payer, wallet.publicKey, null, 6);
    // const mintY = await createMint(connection, wallet.payer, wallet.publicKey, null, 6);
    const mintX=new PublicKey("9QndCCfG6r9d9GnJD6HZZrhzGGRAvSFR3nB3WcyGhVND");
    const mintY=new PublicKey("4uVCs6uDkymcru7hpy4m7E5QtTNBSp2UAbcUEQQti91E");

    const vaultX = getAssociatedTokenAddressSync(
       mintX,
       configPoolPda,
       true // allow PDA owner
    );

    const vaultY = getAssociatedTokenAddressSync(
       mintY,
       configPoolPda,
       true
    );
    console.log("mintX:", mintX.toString())
    console.log("mintY:", mintY.toString())
    console.log("vaultX:", vaultX.toString())
    console.log("vaultY:", vaultY.toString())
    console.log("mintLp:", lpMintPda.toString())
  

    const tx = await program.methods
      .initializePool(
        poolSeed,
        fee,
        authority, // optional: can pass null with `null` if using Option<Pubkey>
      )
      .accounts({
        initializer: provider.publicKey,
        mintX: mintX,
        mintY: mintY,
        configAmm: configAmmPda,
        configPool: configPoolPda,
        mintLp: lpMintPda,
        vaultX,
        vaultY,
        tokenProgram: TOKEN_PROGRAM_ID,
        associatedTokenProgram: ASSOCIATED_TOKEN_PROGRAM_ID,
        systemProgram: SystemProgram.programId,
      } as any)
      .rpc();

    console.log("Pool initialized. Tx:", tx);
  });

  });
// Expected config_pool PDA: 8KroHm9grP1hjAg7CuyNpRFRSAfU5pLktPkruugUsFsz
// Expected LP mint PDA: AGptQGrEs2yQtnFcCpocty6dvkD3jXHLq4jAY4JM6kL1
// mintX: PublicKey(9QndCCfG6r9d9GnJD6HZZrhzGGRAvSFR3nB3WcyGhVND)
// mintY: PublicKey(4uVCs6uDkymcru7hpy4m7E5QtTNBSp2UAbcUEQQti91E)
// vaultX: PublicKey(Be5wK9N2Q51T5s6sCQHTpbf21LWqJ8Zg5ub1j6W42XEY)
// vaultY: PublicKey(7yE3c8TfSSDPsWgzMSUued4QFVVMMJuBi9DoFJeTqt5R)
// mintLp: PublicKey(AGptQGrEs2yQtnFcCpocty6dvkD3jXHLq4jAY4JM6kL1)

File: capstone_launchpad/programs/capstone_launchpad/src/handlers/initialize_pool.rs
use anchor_lang::prelude::*;
use anchor_spl::{associated_token::AssociatedToken, token::{Mint, Token, TokenAccount}};

use crate::state::{ConfigAmm, ConfigPool};

#[derive(Accounts)]
#[instruction(pool_seed: u64)]
pub struct InitializePool<'info> {
    #[account(mut)]
    pub initializer: Signer<'info>,

    pub mint_x: Account<'info, Mint>,
    pub mint_y: Account<'info, Mint>,

        /// Parent AMM config
    #[account(
        mut,
        seeds = [b"config_amm", config_amm.amm_seed.to_le_bytes().as_ref()],
        bump = config_amm.config_amm_bump,
    )]
    pub config_amm: Account<'info, ConfigAmm>,

        /// Pool-specific config
    #[account(
        init,
        payer = initializer,
        seeds = [b"config_pool", config_amm.key().as_ref(), pool_seed.to_le_bytes().as_ref()],
        bump,
        space = 8 + ConfigPool::INIT_SPACE,
    )]
    pub config_pool: Account<'info, ConfigPool>,

    // #[account(
    //     mut,
    //     seeds = [b"config", config.seed.to_le_bytes().as_ref()],
    //     bump = config.config_bump,
    // )]
    // pub config: Account<'info, ConfigPool>,

    #[account(
        init,
        payer = initializer,
        seeds = [b"lp", config_pool.key().as_ref()],
        bump,
        mint::decimals = 6,
        mint::authority = config_pool,
    )]
    pub mint_lp: Account<'info, Mint>,

    #[account(
        init,
        payer = initializer,
        associated_token::mint = mint_x,
        associated_token::authority = config_pool,
    )]
    pub vault_x: Account<'info, TokenAccount>,

    #[account(
        init,
        payer = initializer,
        associated_token::mint = mint_y,
        associated_token::authority = config_pool,
    )]
    pub vault_y: Account<'info, TokenAccount>,

    pub token_program: Program<'info, Token>,
    pub associated_token_program: Program<'info, AssociatedToken>,
    pub system_program: Program<'info, System>,
}

impl<'info> InitializePool<'info> {
    pub fn initialize_pool(
        &mut self,
        pool_seed: u64,
        fee: u16,
        authority: Option<Pubkey>,
        bumps: InitializePoolBumps,
    ) -> Result<()> {
    self.config_pool.set_inner(ConfigPool {
            pool_seed: pool_seed,
            authority,
            mint_x: self.mint_x.key(),
            mint_y: self.mint_y.key(),
            fee,
            locked: false,
            config_pool_bump: bumps.config_pool,
            lp_bump: bumps.mint_lp,
        });
        Ok(())
    }
}

